gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```diff
@@ -0,0 +1,178 @@
+import java.net.SocketAddress;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.ServerSocketChannel;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class implements a tunnel through a proxy to resource on the internet. Frequently data stores to be accessed by
+ * Gobblin reside outside data centers. In these cases, outbound access from a data center typically needs to go through
+ * a gateway proxy for security purposes. In some cases this is an HTTP proxy. However, some protocols like JDBC don't
+ * support the concept of "proxies", let alone HTTP proxies, and hence a solution is needed to enable this.
+ *
+ * This class provides a method of tunneling arbitrary protocls like JDBC connections over an HTTP proxy. Note that
+ * while it is currently only implemented for JDBC (see {@link gobblin.source.extractor.extract.jdbc.JdbcExtractor} and
+ * {@link gobblin.source.extractor.extract.jdbc.JdbcExtractor}), it can be extended to work with any other
+ * TCP-based protocol.
+ *
+ * The way the Tunnel works is as follows:
+ * 1. When a Gobblin data source or Extractor or related class (such as JdbcProvider) is invoked to fetch data from an
+ *    externally hosted resource, it should check if the WorkUnit has a proxy host and port defined.
+ * 2. If a proxy is defined, it should extract the remote host and port from the target URL hosting the resource (e.g.
+ *    the JdbcProvider gets this from the connectionUrl.)
+ * 3. The extractor then creates a Tunnel instance configured with the remote host and port and the proxy host and port.
+ * 4. The Tunnel starts a thread that listens on an arbitrary port on localhost.
+ * 5. The extractor then points the target URL to the localhost and port the Tunnel is listening on. (E.g. in the case
+ *    of JDBC, the JdbcProvider changes the connectionUrl to replace the remote host and port with the localhost and
+ *    port before passing it on to the driver.)
+ * 6. Hence when the extractor client (e.g. JDBC driver) creates a connection, it connects to the Tunnel socket instead
+ *    of the actual target host.
+ * 7. The Tunnel then connects to the remote host through the proxy via a HTTP CONNECT request.
+ * 8. If established successfully, the Tunnel then simply relays bytes back and forth between the Gobblin extractor and
+ *    the target host via the intermediate proxy.)
+ * 7. When the Gobblin extractor (e.g. JDBC data source) is closed down, the Tunnel must be shut down as well.
+ *
+ * The Tunnel can accept as many connections as the JdbcExtractor opens. It uses NIO to minimize resource usage.
+ *
+ * @author navteniev@linkedin.com
+ * @author kkandekar@linkedin.com
+ */
+public class Tunnel {
+  public static final int NON_EXISTENT_PORT = -1;
+
+  private static final Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+
+  private ServerSocketChannel server;
```

Comment:
This is the socket that listens for connections from the client (the JDBC driver), hence it's a server socket. Java distinguishes between server and client sockets, whereas this distinction does not really exist at the C socket layer, where a server socket is identified by the bind and listen calls.